PhysicsTools/HepMCCandAlgos/plugins/GenPUProtonProducer.cc

```diff
@@ -0,0 +1,268 @@
+  abortOnUnknownPDGCode_( cfg.getUntrackedParameter<bool>( "abortOnUnknownPDGCode", true ) ),
+  bunchList_( cfg.getParameter<vector<int> >( "bunchCrossingList" ) ),
+  minPz_( cfg.getParameter<double>( "minPz" ) ) 
+{
+  produces<GenParticleCollection>();
+  mixToken_ = consumes<CrossingFrame<HepMCProduct> >( InputTag(cfg.getParameter<std::string>( "mix" ),"generatorSmeared") );
+}
+
+GenPUProtonProducer::~GenPUProtonProducer() { }
+
+void GenPUProtonProducer::produce( Event& evt, const EventSetup& es ) {
+
+   if ( !convertParticle_.initialized() ) {
+     ESHandle<HepPDT::ParticleDataTable> pdt_handle;
+     es.getData( pdt_handle );
+     HepPDT::ParticleDataTable const& pdt = *pdt_handle.product();
+     convertParticle_.init( pdt ); 
+   }
+
+   unsigned int totalSize = 0;
+   unsigned int npiles = 1;
+
+   Handle<CrossingFrame<HepMCProduct> > cf;
+   evt.getByToken(mixToken_,cf);
+   std::auto_ptr<MixCollection<HepMCProduct> > cfhepmcprod( new MixCollection<HepMCProduct>( cf.product() ) );
+   npiles = cfhepmcprod->size();
+
+   LogDebug("GenPUProtonProducer") << " Number of pile-up events : " << npiles << endl;
+
+   for(unsigned int icf = 0; icf < npiles; ++icf){
+      LogDebug("GenPUProtonProducer") << "CF " << icf << " size : " << cfhepmcprod->getObject(icf).GetEvent()->particles_size() << endl;
+      totalSize += cfhepmcprod->getObject(icf).GetEvent()->particles_size();
+   }
+   LogDebug("GenPUProtonProducer") << "Total size : " << totalSize << endl;
+
+   // Loop over pile-up events
+   MixCollection<HepMCProduct>::MixItr mixHepMC_itr;
+   unsigned int total_number_of_protons = 0;
+   size_t idx_mix = 0;
+   for( mixHepMC_itr = cfhepmcprod->begin() ; mixHepMC_itr != cfhepmcprod->end() ; ++mixHepMC_itr, ++idx_mix ){
+      int bunch = mixHepMC_itr.bunch();
+      if( find( bunchList_.begin(), bunchList_.end(), bunch ) != bunchList_.end() ){
+	 auto event = (*mixHepMC_itr).GetEvent();
+	 // Find protons
+	 unsigned int number_of_protons = 0;
+	 for( auto p = event->particles_begin() ; p != event->particles_end() ; ++p ){
+	    if( select_(*p, minPz_) ) ++number_of_protons;
+	 }
+	 LogDebug("GenPUProtonProducer") << "Idx : " << idx_mix << " Bunch : " << bunch << " Number of protons : " << number_of_protons << endl;
+	 total_number_of_protons += number_of_protons;
+      }
+   }
+   LogDebug("GenPUProtonProducer") << "Total number of protons : " << total_number_of_protons << endl;
+
+   // Initialise containers
+   const size_t size = total_number_of_protons;
+   vector<const HepMC::GenParticle *> particles( size );
+
+   auto candsPtr = std::make_unique<GenParticleCollection>(size);
+   auto barCodeVector = std::make_unique<vector<int>>(size);
+   ref_ = evt.getRefBeforePut<GenParticleCollection>();
+   GenParticleCollection& cands = *candsPtr;
+
+   size_t idx_particle = 0;
+   idx_mix = 0;
+   // Fill collection
+   for( mixHepMC_itr = cfhepmcprod->begin() ; mixHepMC_itr != cfhepmcprod->end() ; ++mixHepMC_itr, ++idx_mix ){
```

Comment:
@mdhildreth Indeed the first loop in the HepMC products could be suppressed. It is reminiscent of the initial structure used as a basis for this module. I added below [1] some time report lines when running the premixing sample with these modules with few events in lxplus. The cmsDriver command is also shown below [2]. The 'genPUProtons*' modules added each around 1.5 ms. Of course the module that dominates the processing time is the mixing module (~30 s in this case).
Please let me know if I should update the module and prepare a new request or a fix could be added later.
Thanks.
[1]
TimeReport ---------- Event  Summary ---[sec]----
TimeReport       event loop CPU/event = 30.289363
TimeReport      event loop Real/event = 52.353183
TimeReport     sum Streams Real/event = 49.928158
TimeReport efficiency CPU/Real/thread = 0.578558
---
TimeReport ---------- Modules in Path: digitisation_step ---[Real sec]----
TimeReport  per event    per visit  Name
TimeReport   0.000000     0.000000  generator
TimeReport  19.516880    19.516880  cpuSpender
TimeReport   0.000031     0.000031  generatorSmeared
TimeReport   0.000584     0.000584  genParticles
TimeReport   0.000225     0.000225  genParticlesForJets
TimeReport   0.000009     0.000009  genParticlesForJetsNoNu
TimeReport   0.000539     0.000539  ak4GenJets
TimeReport   0.000018     0.000018  ak8GenJets
TimeReport   0.000008     0.000008  ak4GenJetsNoNu
TimeReport   0.000006     0.000006  ak8GenJetsNoNu
TimeReport   0.000007     0.000007  genCandidatesForMET
TimeReport   0.000021     0.000021  genParticlesForMETAllVisible
TimeReport   0.000141     0.000141  genMetCalo
TimeReport   0.000006     0.000006  genMetTrue
TimeReport   0.000000     0.000000  randomEngineStateProducer
TimeReport  29.127603    29.127603  mix
TimeReport   0.000000     0.000000  generatorSmeared
TimeReport   0.055196     0.055196  simEcalTriggerPrimitiveDigis
TimeReport   0.020653     0.020653  simEcalDigis
TimeReport   0.000828     0.000828  simHcalDigis
TimeReport   0.142470     0.142470  simMuonCSCDigis
TimeReport   0.000838     0.000838  simMuonDTDigis
TimeReport   0.013858     0.013858  simMuonRPCDigis
TimeReport   0.000217     0.000217  addPileupInfo
TimeReport   0.001565     0.001565  genPUProtonsBx0
TimeReport   0.001450     0.001450  genPUProtonsBxm1
TimeReport   0.001455     0.001455  genPUProtonsBxp1
---
[2]
cmsDriver.py Configuration/Generator/python/SingleNuE10_cfi.py --fileout file:file_out_premixing.root --pileup_input dbs:/RelValProdMinBias_13/CMSSW_9_0_0-90X_mcRun2_asymptotic_v5-v1/GEN-SIM --mc --eventcontent PREMIX --pileup 2016_25ns_Moriond17MC_PoissonOOTPU --datatier GEN-SIM-DIGI-RAW --conditions 90X_mcRun2_asymptotic_v5 --step GEN,SIM,DIGIPREMIX,L1,DIGI2RAW --era Run2_2016 -n 30 --customise PhysicsTools/HepMCCandAlgos/addGenPUProtons.customiseGenPUProtons,PhysicsTools/HepMCCandAlgos/addGenPUProtonsEventContent.customiseGenPUProtonsEventContent

---